File: iphone/ButtonLog/Models/Support.swift
import Foundation

// MARK: - Support Ticket

struct SupportTicket: Identifiable, Codable {
    let id: String
    let subject: String
    let category: TicketCategory
    let priority: TicketPriority
    let status: TicketStatus
    let unreadCount: Int?
    let assignedAdmin: SupportAdmin?
    let createdAt: Date
    let updatedAt: Date
    var messages: [TicketMessage]?

    enum CodingKeys: String, CodingKey {
        case id, subject, category, priority, status, messages
        case unreadCount = "unread_count"
        case assignedAdmin = "assigned_admin"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Ticket Message

struct TicketMessage: Identifiable, Codable {
    let id: String
    let content: String
    let senderId: String
    let senderName: String?
    let isFromSupport: Bool
    let createdAt: Date
    let readAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, content
        case senderId = "sender_id"
        case senderName = "sender_name"
        case isFromSupport = "is_from_support"
        case createdAt = "created_at"
        case readAt = "read_at"
    }
}

// MARK: - Support Admin

struct SupportAdmin: Codable {
    let id: String
    let name: String?
}

// MARK: - Enums

enum TicketCategory: String, Codable, CaseIterable {
    case bug = "bug"
    case featureRequest = "feature_request"
    case question = "question"
    case other = "other"

    var displayName: String {
        switch self {
        case .bug: return "Bug Report"
        case .featureRequest: return "Feature Request"
        case .question: return "Question"
        case .other: return "Other"
        }
    }

    var icon: String {
        switch self {
        case .bug: return "ladybug"
        case .featureRequest: return "lightbulb"
        case .question: return "questionmark.circle"
        case .other: return "ellipsis.circle"
        }
    }
}

enum TicketPriority: String, Codable, CaseIterable {
    case low = "low"
    case normal = "normal"
    case high = "high"
    case urgent = "urgent"

    var displayName: String {
        switch self {
        case .low: return "Low"
        case .normal: return "Normal"
        case .high: return "High"
        case .urgent: return "Urgent"
        }
    }
}

enum TicketStatus: String, Codable, CaseIterable {
    case open = "open"
    case inProgress = "in_progress"
    case resolved = "resolved"
    case closed = "closed"

    var displayName: String {
        switch self {
        case .open: return "Open"
        case .inProgress: return "In Progress"
        case .resolved: return "Resolved"
        case .closed: return "Closed"
        }
    }

    var color: String {
        switch self {
        case .open: return "yellow"
        case .inProgress: return "blue"
        case .resolved: return "green"
        case .closed: return "gray"
        }
    }

    var isActive: Bool {
        switch self {
        case .open, .inProgress: return true
        case .resolved, .closed: return false
        }
    }
}

// MARK: - Form Data

struct TicketFormData {
    var subject: String = ""
    var category: TicketCategory = .question
    var priority: TicketPriority = .normal
    var message: String = ""

    var isValid: Bool {
        !subject.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func toRequestBody() -> [String: Any] {
        return [
            "ticket": [
                "subject": subject.trimmingCharacters(in: .whitespacesAndNewlines),
                "category": category.rawValue,
                "priority": priority.rawValue,
                "message": message.trimmingCharacters(in: .whitespacesAndNewlines)
            ]
        ]
    }
}

struct MessageFormData {
    var content: String = ""

    var isValid: Bool {
        !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func toRequestBody() -> [String: Any] {
        return [
            "message": [
                "content": content.trimmingCharacters(in: .whitespacesAndNewlines)
            ]
        ]
    }
}

File: android/app/src/main/java/com/buttonlog/app/data/model/Support.kt
package com.buttonlog.app.data.model

import androidx.compose.ui.graphics.Color
import com.google.gson.annotations.SerializedName

// MARK: - Support Ticket

data class SupportTicket(
    val id: String,
    val subject: String,
    val category: TicketCategory,
    val priority: TicketPriority,
    val status: TicketStatus,
    @SerializedName("unread_count")
    val unreadCount: Int?,
    @SerializedName("assigned_admin")
    val assignedAdmin: SupportAdmin?,
    @SerializedName("created_at")
    val createdAt: String,
    @SerializedName("updated_at")
    val updatedAt: String,
    val messages: List<TicketMessage>?
)

// MARK: - Ticket Message

data class TicketMessage(
    val id: String,
    val content: String,
    @SerializedName("sender_id")
    val senderId: String,
    @SerializedName("sender_name")
    val senderName: String?,
    @SerializedName("is_from_support")
    val isFromSupport: Boolean,
    @SerializedName("created_at")
    val createdAt: String,
    @SerializedName("read_at")
    val readAt: String?
)

// MARK: - Support Admin

data class SupportAdmin(
    val id: String,
    val name: String?
)

// MARK: - Enums

enum class TicketCategory(val displayName: String, val icon: String) {
    @SerializedName("bug")
    BUG("Bug Report", "bug_report"),

    @SerializedName("feature_request")
    FEATURE_REQUEST("Feature Request", "lightbulb"),

    @SerializedName("question")
    QUESTION("Question", "help"),

    @SerializedName("other")
    OTHER("Other", "more_horiz")
}

enum class TicketPriority(val displayName: String) {
    @SerializedName("low")
    LOW("Low"),

    @SerializedName("normal")
    NORMAL("Normal"),

    @SerializedName("high")
    HIGH("High"),

    @SerializedName("urgent")
    URGENT("Urgent")
}

enum class TicketStatus(val displayName: String, val color: Color) {
    @SerializedName("open")
    OPEN("Open", Color(0xFFFF9800)),

    @SerializedName("in_progress")
    IN_PROGRESS("In Progress", Color(0xFF2196F3)),

    @SerializedName("resolved")
    RESOLVED("Resolved", Color(0xFF4CAF50)),

    @SerializedName("closed")
    CLOSED("Closed", Color(0xFF9E9E9E));

    val isActive: Boolean
        get() = this == OPEN || this == IN_PROGRESS
}

// MARK: - Form Data

data class TicketFormData(
    var subject: String = "",
    var category: TicketCategory = TicketCategory.QUESTION,
    var priority: TicketPriority = TicketPriority.NORMAL,
    var message: String = ""
) {
    val isValid: Boolean
        get() = subject.trim().isNotEmpty() && message.trim().isNotEmpty()
}

// MARK: - API Request/Response

data class CreateTicketRequest(
    val ticket: TicketData
)

data class TicketData(
    val subject: String,
    val category: String,
    val priority: String,
    val message: String
)

data class SendMessageRequest(
    val message: MessageData
)

data class MessageData(
    val content: String
)

// API response wrappers
data class SupportTicketsResponse(
    val success: Boolean,
    val data: List<SupportTicket>?,
    val error: ApiError?,
    val meta: ApiMeta?
)

data class SupportTicketResponse(
    val success: Boolean,
    val data: SupportTicket?,
    val error: ApiError?,
    val meta: ApiMeta?
)

data class TicketMessageResponse(
    val success: Boolean,
    val data: TicketMessage?,
    val error: ApiError?,
    val meta: ApiMeta?
)
